# Web Scraping Aassignment 1

In [1]:
# Importing required libraries
from bs4 import BeautifulSoup
import requests

## 1. Write a python program to display all the header tags from ‘en.wikipedia.org/wiki/Main_Page’.

In [2]:
# Function to scrape all headers
def sc_header(url = None):
    page = requests.get(url)
    print(page)
    data = page.text
    soup = BeautifulSoup(data, 'lxml')
    first_header = soup.find_all([f'h{i}' for i in range(1,10) ])
    headers = []
    for i in first_header:
        headers.append(i.text.replace('\n',""))
        
    df = pd.DataFrame({})
    df['headers'] = headers[:]
    return(df.to_csv('Project_1.csv'))
# Calling the function
sc_header(url = "http://en.wikipedia.org/wiki/Main_Page")

<Response [200]>


<IPython.core.display.Javascript object>

In [61]:
header = pd.read_csv('Project_1.csv')
header.drop('Unnamed: 0',axis = 1)

<IPython.core.display.Javascript object>

headers
0                       Main Page
1   From today's featured article
2                Did you know ...
3                     In the news
4                     On this day
5      From today's featured list
6        Today's featured picture
7        Other areas of Wikipedia
8     Wikipedia's sister projects
9             Wikipedia languages
10                Navigation menu
11                 Personal tools
12                     Namespaces
13                       Variants
14                          Views
15                           More
16                         Search
17                     Navigation
18                     Contribute
19                          Tools
20                   Print/export
21              In other projects
22                      Languages

## 2. Write a python program to display IMDB’s Top rated 100 movies’ data (i.e. Name, IMDB rating, Year of release) and save it in form of a CSV file. 

In [56]:
# Function to scrape top rated Movie names, IMDB ratings & Year of release
def movie_scrape(url = None):
    page = requests.get(url)
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    r = soup.find_all('td', class_='ratingColumn imdbRating')
    m_name = [x.a.text for x in soup.find_all('td', class_ = "titleColumn")]
    y = [x.text for x in soup.find_all('span', class_='secondaryInfo')]
    name = []
    rating = []
    year = []
    for i in m_name:
        name.append(i.replace('\n',""))
    for j in y:
        year.append(j)
    for k in r:
        rating.append(k.text.replace('\n',""))
    df = pd.DataFrame({})
    df['Movie_Name'] = name[:100]
    df['Year of Release']= year[:100]
    df['Year of Release'] = df['Year of Release'].str.extract('(\d+)')
    df['Rating'] = rating[:100]
    return(df.to_csv('Project_2.csv'))
    
# Calling the function
movie_scrape(url = "https://www.imdb.com/chart/top/?ref_=nv_mv_250")

<IPython.core.display.Javascript object>

In [62]:
top_movies = pd.read_csv('Project_2.csv')
top_movies.drop('Unnamed: 0',axis = 1)

<IPython.core.display.Javascript object>

Movie_Name  Year of Release  Rating
0   The Shawshank Redemption             1994     9.2
1              The Godfather             1972     9.1
2     The Godfather: Part II             1974     9.0
3            The Dark Knight             2008     9.0
4               12 Angry Men             1957     8.9
..                       ...              ...     ...
95              Citizen Kane             1941     8.3
96                    Dangal             2016     8.3
97       Singin' in the Rain             1952     8.2
98       Ladri di biciclette             1948     8.2
99                   The Kid             1921     8.2

[100 rows x 3 columns]

## 3. Write a python program to display IMDB’s Top rated 100 Indian movies’ data (i.e. Name, IMDB rating, Year of release) and save it in form of a CSV file.

In [58]:
# Function to scrape top rated Indian Movie names, IMDB ratings & Year of release
def movie_scrape(url = None):
    page = requests.get(url)
    data = page.text
    soup = BeautifulSoup(data, 'lxml')
    r = soup.find_all('td', class_='ratingColumn imdbRating')
    m_name = [x.a.text for x in soup.find_all('td', class_ = "titleColumn")]
    y = soup.find_all('span', class_='secondaryInfo')
    name = []
    rating = []
    year = []
    for i in m_name:
        name.append(i.replace('\n',""))
    for j in y:
        year.append(j.text.replace('\n',""))
    for k in r:
        rating.append(k.text.replace('\n',""))
    df = pd.DataFrame({})
    df['Movie_Name'] = name[:100]
    df['Year of Release']= year[:100]
    df['Year of Release'] = df['Year of Release'].str.extract('(\d+)')
    df['Rating'] = rating[:100]
    return(df.to_csv('Project_3.csv'))
# Calling the function
movie_scrape(url = "https://www.imdb.com/india/top-rated-indian-movies/")

<IPython.core.display.Javascript object>

In [63]:
hindi_movies = pd.read_csv('Project_3.csv')
hindi_movies.drop('Unnamed: 0',axis = 1)

<IPython.core.display.Javascript object>

Movie_Name  Year of Release  Rating
0              Pather Panchali             1955     8.5
1                     Gol Maal             1979     8.5
2                      Nayakan             1987     8.5
3                   Anbe Sivam             2003     8.5
4                   Drishyam 2             2021     8.5
..                         ...              ...     ...
95                      Barfi!             2012     8.0
96  The Legend of Bhagat Singh             2002     8.0
97                  Bommarillu             2006     8.0
98                     Maqbool             2003     8.0
99                      Bombay             1995     8.0

[100 rows x 3 columns]

## 4. Write a python program to scrap book name, author name, genre and book review of any 5 books from‘www.bookpage.com’

In [72]:
def book_scrape(url = None):
    page = requests.get(url)
    data = page.text
    soup = BeautifulSoup(data, 'lxml')
    book_names = [x.a.text for x in soup.find_all('h4', class_ = "italic")]
    author_names = [x.text for x in soup.find_all('p', class_ = "sans bold")]
    genre = [x.text for x in soup.find_all('p', class_ = "genre-links hidden-phone")]
    review = [x.text for x in soup.find_all('p', class_ = "excerpt")]
    df = pd.DataFrame({})
    df['book_names'] = book_names[:5]
    df['author_names']= author_names[:5]
    df['genre'] = genre[:5]
    df['review'] = review[:5]
    return(df.to_csv('Project_4.csv'))

book_scrape(url = "https://bookpage.com/reviews")

<IPython.core.display.Javascript object>

In [76]:
books = pd.read_csv('Project_4.csv')
books.drop('Unnamed: 0',axis = 1,inplace = True)
books['book_names']=books['book_names'].str.replace('★',"")
books['author_names']=books['author_names'].str.replace('\n',"")
books['genre']=books['genre'].str.replace('\n',"")
books['review']=books['review'].str.replace('\n',"")
books

<IPython.core.display.Javascript object>

book_names                     author_names  \
0                       Just as I Am        Cicely Tyson, Robin Miles   
1                   Red Island House                       Andrea Lee   
2    Flight of the Diamond Smugglers              Matthew Gavin Frank   
3           The One Thing You'd Save  Linda Sue Park, Robert Sae-Heng   
4            The Daughters of Kobani             Gayle Tzemach Lemmon   

                                           genre  \
0        Audio / Nonfiction / Biography & Memoir   
1                     Fiction / Literary Fiction   
2                  Nonfiction / Memoir / Animals   
3                      Children's / Middle Grade   
4  Nonfiction / History / Middle Eastern History   

                                              review  
0  Narrator Robin Miles bringing the same warmth ...  
1  The stories of Andrea Lee’s Red Island House a...  
2  Matthew Gavin Frank's Flight of the Diamond Sm...  
3                                                     
4  The story of how young Kurdish women brought d...

## 5. Write a python program to scrape cricket rankings from ‘www.icc-cricket.com’. You have to scrape: 

### i) Top 10 ODI teams in men’s cricket along with the records for matches, points and rating.

In [6]:
def odi_team(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['matches'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    return(df.to_csv('Project_5_1.csv'))


odi_team(url = 'https://www.icc-cricket.com/rankings/mens/team-rankings/odi')

<IPython.core.display.Javascript object>

In [88]:
odi_team = pd.read_csv('Project_5_1.csv')
odi_team.drop('Unnamed: 0',axis = 1,inplace = True)
odi_team.head(10)

<IPython.core.display.Javascript object>

ranks           names  matches points  rating
0      1      EnglandENG       44  5,405     123
1      2        IndiaIND       52  6,102     117
2      3   New ZealandNZ       33  3,857     117
3      4    AustraliaAUS       39  4,344     111
4      5  South AfricaSA       31  3,345     108
5      6     PakistanPAK       35  3,490     100
6      7   BangladeshBAN       38  3,432      90
7      8     Sri LankaSL       42  3,372      80
8      9   West IndiesWI       49  3,802      78
9     10  AfghanistanAFG       31  1,844      59

### ii) Top 10 ODI Batsmen in men along with the records of their team and rating.

In [91]:
def odi_batsmen(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['team'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    df['rating'] = df['rating'].str.extract('(\d+)')
    return(df.to_csv('Project_5_2.csv'))


odi_batsmen(url = 'https://www.icc-cricket.com/rankings/mens/player-rankings/odi/batting')

<IPython.core.display.Javascript object>

In [92]:
odi_batsmen = pd.read_csv('Project_5_2.csv')
odi_batsmen.drop('Unnamed: 0',axis = 1,inplace = True)
odi_batsmen.ranks.replace(np.NaN,8,inplace = True)
odi_batsmen.head(10)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

ranks                names team  points  \
0                                       1          Virat Kohli  IND     870   
1                               2                 Rohit Sharma  IND     842   
2                               3                   Babar Azam  PAK     837   
3                               4                  Ross Taylor   NZ     818   
4                               5                  Aaron Finch  AUS     791   
5                               6          Francois du Plessis   SA     790   
6                               7                 David Warner  AUS     773   
7                                       8            Shai Hope   WI     773   
8                               9              Kane Williamson   NZ     765   
9                              10              Quinton de Kock   SA     755   

   rating  
0     911  
1     885  
2     846  
3     841  
4     798  
5     820  
6     880  
7     808  
8     799  
9     813

### iii) Top 10 ODI bowlers along with the records of their team and rating.

In [93]:
def odi_bowlers(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['team'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    df['rating'] = df['rating'].str.extract('(\d+)')
    return(df.to_csv('Project_5_3.csv'))


odi_bowlers(url = 'https://www.icc-cricket.com/rankings/mens/player-rankings/odi/bowling')

<IPython.core.display.Javascript object>

In [95]:
odi_bowler = pd.read_csv('Project_5_3.csv')
odi_bowler.drop('Unnamed: 0',axis = 1,inplace = True)
odi_bowler.head(10)

<IPython.core.display.Javascript object>

ranks              names team  points  \
0                                       1        Trent Boult   NZ     722   
1                               2           Mujeeb Ur Rahman  AFG     708   
2                               3             Jasprit Bumrah  IND     700   
3                               4               Mehedi Hasan  BAN     694   
4                               5               Chris Woakes  ENG     675   
5                               6              Kagiso Rabada   SA     665   
6                               7             Josh Hazlewood  AUS     660   
7                               8          Mustafizur Rahman  BAN     658   
8                               9              Mohammad Amir  PAK     647   
9                              10                Pat Cummins  AUS     646   

   rating  
0   770.0  
1   712.0  
2   841.0  
3   694.0  
4   676.0  
5   724.0  
6   733.0  
7   695.0  
8   663.0  
9   729.0

## 6. Write a python program to scrape cricket rankings from ‘www.icc-cricket.com’. You have to scrape:
### i) Top 10 ODI teams in women’s cricket along with the records for matches, points and rating. 

In [9]:
def odi_team_women(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['matches'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    return(df.to_csv('Project_6_1.csv'))


odi_team_women(url = 'https://www.icc-cricket.com/rankings/womens/team-rankings/odi')

<IPython.core.display.Javascript object>

In [96]:
odi_team_women = pd.read_csv('Project_6_1.csv')
odi_team_women.drop('Unnamed: 0',axis = 1,inplace = True)
odi_team_women

<IPython.core.display.Javascript object>

ranks           names  matches points  rating
0      1    AustraliaAUS       15  2,436     162
1      2  South AfricaSA       24  2,828     118
2      3      EnglandENG       17  1,993     117
3      4        IndiaIND       20  2,226     111
4      5   New ZealandNZ       18  1,696      94
5      6   West IndiesWI       12  1,025      85
6      7     PakistanPAK       15  1,101      73
7      8   BangladeshBAN        5    306      61
8      9     Sri LankaSL       11    519      47
9     10      IrelandIRE        2     25      13

### ii) Top 10 women’s ODI players along with the records of their team and rating. 

In [84]:
def odi_player_women(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['matches'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    df['rating'] = df['rating'].str.extract('(\d+)')
    df['ranks'] = df['ranks'].str.extract('(\d+)')
    return(df.to_csv('Project_6_2.csv'))


odi_player_women(url = 'https://www.icc-cricket.com/rankings/womens/player-rankings/odi/batting')

<IPython.core.display.Javascript object>

In [97]:
odi_women_player = pd.read_csv('Project_6_2.csv')
odi_women_player.drop('Unnamed: 0',axis = 1,inplace = True)
odi_women_player.head(10)

<IPython.core.display.Javascript object>

ranks              names matches  points  rating
0    1.0        Lizelle Lee      SA     773     773
1    2.0     Tammy Beaumont     ENG     765     765
2    3.0        Meg Lanning     AUS     749     834
3    4.0    Stafanie Taylor      WI     746     765
4    5.0       Alyssa Healy     AUS     741     741
5    6.0  Amy Satterthwaite      NZ     740     756
6    7.0    Smriti Mandhana     IND     719     797
7    8.0    Laura Wolvaardt      SA     699     725
8    9.0        Mithali Raj     IND     693     839
9   10.0       Ellyse Perry     AUS     691     766

### iii) Top 10 women’s ODI all-rounder along with the records of their team and rating.

In [86]:
def odi_allrounder_women(url = None):
    page = requests.get(url )
    data = page.text
    soup = BeautifulSoup(data, 'html.parser')
    teams = soup.find_all('td')
    team_names = []
    for i in teams:
        team_names.append(i.text.replace('\n',""))
    df = pd.DataFrame({})
    df['ranks'] = team_names[0::5]
    df['names'] = team_names[1::5]
    df['matches'] = team_names[2::5]
    df['points'] = team_names[3::5]
    df['rating'] = team_names[4::5]
    df['rating'] = df['rating'].str.extract('(\d+)')
    return(df.to_csv('Project_6_3.csv'))


odi_allrounder_women(url = 'https://www.icc-cricket.com/rankings/womens/player-rankings/odi/all-rounder')

<IPython.core.display.Javascript object>

In [98]:
odi_ar_women = pd.read_csv('Project_6_3.csv')
odi_ar_women.drop('Unnamed: 0',axis = 1,inplace = True)
odi_ar_women.head(10)

<IPython.core.display.Javascript object>

ranks             names matches  points  \
0                                       1      Ellyse Perry     AUS     460   
1                               2           Stafanie Taylor      WI     410   
2                               3            Marizanne Kapp      SA     390   
3                               4             Deepti Sharma     IND     357   
4                               5            Natalie Sciver     ENG     349   
5                               6             Jess Jonassen     AUS     301   
6                               7             Sophie Devine      NZ     274   
7                               8          Dane van Niekerk      SA     252   
8                               9           Katherine Brunt     ENG     236   
9                              10          Ashleigh Gardner     AUS     223   

   rating  
0     548  
1     559  
2     412  
3     397  
4     349  
5     308  
6     305  
7     421  
8     270  
9     241

## 7. Write a python program to scrape details of all the mobile phones under Rs. 20,000 listed on Amazon.in. The scraped data should include Product Name, Price, Image URL and Average Rating.

In [12]:
page = requests.get(url= 'https://www.amazon.in/s?bbn=1389401031&rh=n%3A1389401031%2Cp_36%3A1318506031&dc&qid=1616327014&rnid=1318502031&ref=lp_1389401031_nr_p_36_3')

In [13]:
page

<Response [503]>

Scraping was not accepted by Amazon India. Status code 503 means that the server is down for maintenance.

## 8. Write a python program to extract information about the local weather from the National Weather Service website of USA, https://www.weather.gov/ for the city, San Francisco. You need to extract data about 7 day extended forecast display for the city. The data should include period, short description, temperature and description. 

In [14]:
def weather_report(url = None):
    page = requests.get(url)
    data = page.text
    soup = BeautifulSoup(data, 'lxml')
    Period = [x.text for x in soup.find_all('p', class_="period-name")]
    Short_desc = [x.text for x in soup.find_all('p', class_="short-desc")]
    Temp_high = [x.text for x in soup.find_all('p', class_="temp temp-high")]
    Temp_Low = [x.text for x in soup.find_all('p', class_="temp temp-low")]
    Desc = [x.text for x in soup.find_all('div', class_="col-sm-10 forecast-text")]
    def twolists(list1, list2):
        newlist = []
        a1 = len(list1)
        a2 = len(list2)
        for i in range(max(a1, a2)):
            if i < a1:
                newlist.append(list1[i])
            if i < a2:
                newlist.append(list2[i])

        return newlist

    temp = twolists(Temp_high,Temp_Low)
    df = pd.DataFrame({})
    df['Period'] = Period
    df['Short_desc'] = Short_desc
    df['temp'] = temp
    df['Long_Desc'] = Desc[0:9]
    
    return(df.to_csv('Project_8.csv'))


weather_report(url= 'https://forecast.weather.gov/MapClick.php?CityName=San+Francisco&state=CA&site=MTR&textField1=37.775&textField2=-122.418#.YFeTkXtR3IU')

<IPython.core.display.Javascript object>

In [99]:
whtr = pd.read_csv('Project_8.csv')
whtr.drop('Unnamed: 0',axis = 1,inplace = True)
whtr

<IPython.core.display.Javascript object>

Period    Short_desc         temp  \
0       Overnight  Mostly Clear  High: 65 °F   
1          Monday  Mostly Sunny   Low: 46 °F   
2     MondayNight  Mostly Clear  High: 69 °F   
3         Tuesday         Sunny   Low: 47 °F   
4    TuesdayNight         Clear  High: 70 °F   
5       Wednesday         Sunny   Low: 49 °F   
6  WednesdayNight  Mostly Clear  High: 65 °F   
7        Thursday         Sunny   Low: 49 °F   
8   ThursdayNight         Clear   Low: 46 °F   

                                           Long_Desc  
0  Mostly clear, with a low around 46. West wind ...  
1  Mostly sunny, with a high near 65. West wind 5...  
2  Mostly clear, with a low around 47. West north...  
3  Sunny, with a high near 69. North wind 7 to 12...  
4  Clear, with a low around 49. North wind 13 to ...  
5                        Sunny, with a high near 70.  
6                Mostly clear, with a low around 49.  
7                        Sunny, with a high near 65.  
8                       Clear, with a low around 46.

## 9. Write a python program to scrape ‘software developer’ job listings from ‘Monster.com’. It should include all the jobs listed for the next 5 pages. 

In [32]:
requests.get(url = 'https://www.monstergulf.com/srp/results?query=Software%20Developer,software%20development&searchId=721946ce-6082-455c-9442-ec2c0bccba86')

<Response [403]>

Website dosent allow data scraping but loop can be used to scrape data from 5 subsequent pages

## 10. Write a python program to scrape ‘data scientist’ job listings for location ‘New Delhi’ from ‘Monster.com’. It should include all the jobs listed for the next 5 pages

In [33]:
requests.get(url='https://www.monsterindia.com/srp/results?query=Data%20Scientist&locations=Delhi&searchId=5e30757d-9dd9-46ee-9e46-8f6035fe76ef')

<Response [403]>

Website dosent allow data scraping but loop can be used to scrape data from 5 subsequent pages